src/Shared/FileSystem/ManagedFileSystem.cs

```diff
@@ -38,19 +39,64 @@ public byte[] ReadFileAllBytes(string path)
             return File.ReadAllBytes(path);
         }
 
+#if FEATURE_MSIOREDIST
+        private IEnumerable<string> HandleFileLoadException(
+            Func<string, string, Microsoft.IO.SearchOption, IEnumerable<string>> enumerateFunctionDelegate,
+            string path,
+            string searchPattern,
+            Microsoft.IO.SearchOption searchOption
+        )
+        {
+            try
+            {
+                return enumerateFunctionDelegate(path, searchPattern, searchOption);
+            }
+            // Microsoft.IO.Redist has a dependency on System.Buffers and if System.Buffers assembly is not found the line above throws an exception.
+            // However, FileMatcher class (that in most cases calls the enumeration) does not allow to fail on a IO-related exception. Such behavior hides the actual exception and makes it obscure.
+            // We rethrow it to make it fail with a proper error message and call stack.
+            catch (FileLoadException ex)
+            {
+                throw new InvalidOperationException("Could not load file or assembly.", ex);
+            }
+            // Sometimes FileNotFoundException is thrown when there is an assembly load failure. In this case it should have FusionLog.
+            catch (FileNotFoundException ex) when (ex.FusionLog != null)
+            {
+                throw new InvalidOperationException("Could not load file or assembly.", ex);
+            }
+        }
+#endif
+
         public virtual IEnumerable<string> EnumerateFiles(string path, string searchPattern, SearchOption searchOption)
         {
+#if FEATURE_MSIOREDIST
+            return ChangeWaves.AreFeaturesEnabled(ChangeWaves.Wave17_0)
+                    ? HandleFileLoadException(Microsoft.IO.Directory.EnumerateFiles, path, searchPattern, (Microsoft.IO.SearchOption)searchOption)
```

Comment:
This creates a new `Func` delegate on each call, a small but avoidable allocation.

![image](https://user-images.githubusercontent.com/12206368/135832141-f5777a49-f7bd-4eba-a048-a3ff38b9c31c.png)

You should be able to get C# cache the delegate if you use a lambda:

```suggestion
                ? HandleFileLoadException((path, searchPattern, searchOption) => Microsoft.IO.Directory.EnumerateDirectories(path, searchPattern, searchOption), path, searchPattern, (Microsoft.IO.SearchOption)searchOption)
```

or pass an enum to `HandleFileLoadException` and use a switch when making the actual call.